    > Version 1.0 of the attribute classification exercise

In [2]:
import mysql.connector
import pandas as pd
mydb=mysql.connector.connect(host='report-reader.cluster-custom-c8oe0gvszktr.ap-south-1.rds.amazonaws.com',
                             user='skreadonlyuser_api',
                             password='XUerT9JxnWnEstSA')
cursor=mydb.cursor()

In [3]:
cursor.execute('''
Select 
v.variantId as VariantId,
upper(clr.colorName) as Color,
clr.colorHex as ColorHex,
case when Upper(cm.name) = 'T-SHIRT' then 'T-SHIRTS' else cm.name end as Category,
cm1.name as MainCategory,
cm2.name as SuperCategory,
pa.attribute as ProductAttribute
from shoekonnect_live.variants as v 
INNER JOIN shoekonnect_live.colors as clr on clr.uniqueId = v.colorId
INNER JOIN shoekonnect_live.products as p on p.productid = v.productid 
INNER JOIN shoekonnect_live.category_master as cm on cm.categoryid = p.subcategoryid
INNER JOIN shoekonnect_live.category_master as cm1 on cm1.categoryid = cm.parentid
INNER JOIN shoekonnect_live.category_master as cm2 on cm2.categoryid = cm1.parentid
INNER JOIN shoekonnect_live.products_attribute as pa on pa.productid = p.productid
where upper(cm.name) IN ('JEANS', 'T-SHIRTS', 'T-SHIRT') and pa.attribute IS NOT NULL
''')

In [4]:
import pandas as pd
import json
rows=cursor.fetchall()
columns=[i[0] for i in  cursor.description]
df=pd.DataFrame(rows,columns=columns)
df

VariantId                                   Color ColorHex  Category  \
0        1513399                                   CREAM  #FFFFCC  t-shirts   
1        1530083  BLACK-1, GOLDEN-1, RED-1, ROYAL BLUE-1           t-shirts   
2        1547548                                   KHAKI  #a18f5e     jeans   
3        1547567                                    BLUE  #0083C3     jeans   
4        1569064                            MULTICOLOR-5           t-shirts   
...          ...                                     ...      ...       ...   
59187    2652696                                SKY BLUE  #74D1EA     jeans   
59188    2652713                                    GREY  #adadad     jeans   
59189    2652715                                SKY BLUE  #74D1EA     jeans   
59190    2652716                                    BLUE  #0083C3     jeans   
59191    2652717                               DARK GREY  #545454     jeans   

      MainCategory SuperCategory  \
0              Men      Apparels   
1              Men      Apparels   
2      Boys - Kids      Apparels   
3      Boys - Kids      Apparels   
4      Boys - Kids      Apparels   
...            ...           ...   
59187        Women      Apparels   
59188  Boys - Kids      Apparels   
59189          Men      Apparels   
59190          Men      Apparels   
59191          Men      Apparels   

                                        ProductAttribute  
0      [{"name": "Fabric", "value": "Cotton", "valueI...  
1      [{"name": "Fabric", "value": "Polyester", "val...  
2      [{"name": "Style", "value": "Embroidered Jeans...  
3      [{"name": "Style", "value": "Embroidered Jeans...  
4      [{"name": "Style", "value": "Solid / Plain T-S...  
...                                                  ...  
59187  [{"name": "fabric", "value": "Cotton Poly 2/1 ...  
59188  [{"name": "fabric", "value": "Cotton Lycra Kni...  
59189  [{"name": "wash quality", "value": "enzyme wit...  
59190  [{"name": "wash quality", "value": "enzyme wit...  
59191  [{"name": "wash quality", "value": "enzyme wit...  

[59192 rows x 7 columns]

In [5]:
df

VariantId                                   Color ColorHex  Category  \
0        1513399                                   CREAM  #FFFFCC  t-shirts   
1        1530083  BLACK-1, GOLDEN-1, RED-1, ROYAL BLUE-1           t-shirts   
2        1547548                                   KHAKI  #a18f5e     jeans   
3        1547567                                    BLUE  #0083C3     jeans   
4        1569064                            MULTICOLOR-5           t-shirts   
...          ...                                     ...      ...       ...   
59187    2652696                                SKY BLUE  #74D1EA     jeans   
59188    2652713                                    GREY  #adadad     jeans   
59189    2652715                                SKY BLUE  #74D1EA     jeans   
59190    2652716                                    BLUE  #0083C3     jeans   
59191    2652717                               DARK GREY  #545454     jeans   

      MainCategory SuperCategory  \
0              Men      Apparels   
1              Men      Apparels   
2      Boys - Kids      Apparels   
3      Boys - Kids      Apparels   
4      Boys - Kids      Apparels   
...            ...           ...   
59187        Women      Apparels   
59188  Boys - Kids      Apparels   
59189          Men      Apparels   
59190          Men      Apparels   
59191          Men      Apparels   

                                        ProductAttribute  
0      [{"name": "Fabric", "value": "Cotton", "valueI...  
1      [{"name": "Fabric", "value": "Polyester", "val...  
2      [{"name": "Style", "value": "Embroidered Jeans...  
3      [{"name": "Style", "value": "Embroidered Jeans...  
4      [{"name": "Style", "value": "Solid / Plain T-S...  
...                                                  ...  
59187  [{"name": "fabric", "value": "Cotton Poly 2/1 ...  
59188  [{"name": "fabric", "value": "Cotton Lycra Kni...  
59189  [{"name": "wash quality", "value": "enzyme wit...  
59190  [{"name": "wash quality", "value": "enzyme wit...  
59191  [{"name": "wash quality", "value": "enzyme wit...  

[59192 rows x 7 columns]

In [6]:
# Function to process each chunk of data
def process_data(df):
    # Parse the JSON in 'ProductAttribute'
    df['ProductAttribute'] = df['ProductAttribute'].apply(json.loads)

    # Explode the 'ProductAttribute' column so each list element becomes its own row
    df_exploded = df.explode('ProductAttribute')

    # Normalize the JSON structure into a dataframe
    df_normalized = pd.json_normalize(df_exploded['ProductAttribute'])

    # Concatenate with the original 'VariantId' column
    df_final = pd.concat([df_exploded['VariantId'].reset_index(drop=True), df_normalized[['name', 'value']]], axis=1)

    # Pivot the table to get 'name' as columns and 'value' as data points
    df_pivot = df_final.pivot_table(index='VariantId', columns='name', values='value', aggfunc='first').reset_index()

    return df_pivot

# Process the data
df_result = process_data(df)

# Show the result
df_result

name   VariantId Belt Type Buttons Closure Design     Fabric  \
0        1513399       NaN     NaN     NaN    NaN     Cotton   
1        1530083       NaN     NaN     NaN    NaN  Polyester   
2        1547548       NaN     NaN     NaN    NaN      Denim   
3        1547567       NaN     NaN     NaN    NaN      Denim   
4        1569064       NaN     NaN     NaN    NaN    Hosiery   
...          ...       ...     ...     ...    ...        ...   
59187    2652696       NaN     NaN     NaN    NaN        NaN   
59188    2652713       NaN     NaN     NaN    NaN        NaN   
59189    2652715       NaN     NaN     NaN    NaN        NaN   
59190    2652716       NaN     NaN     NaN    NaN        NaN   
59191    2652717       NaN     NaN     NaN    NaN        NaN   

name  Fabric Composition Fabric Grade Fabric Weight          Fit  ...  \
0                    NaN      200 GSM           NaN          NaN  ...   
1                    NaN          NaN           NaN          NaN  ...   
2                    NaN          NaN           NaN  Regular Fit  ...   
3                    NaN          NaN           NaN  Regular Fit  ...   
4                    NaN          NaN       180 GSM          NaN  ...   
...                  ...          ...           ...          ...  ...   
59187                NaN          NaN           NaN          NaN  ...   
59188                NaN          NaN           NaN          NaN  ...   
59189                NaN          NaN           NaN          NaN  ...   
59190                NaN          NaN           NaN          NaN  ...   
59191                NaN          NaN           NaN          NaN  ...   

name    softner stiching quality strechability      stretch style top fabric  \
0           NaN              NaN           NaN          NaN   NaN        NaN   
1           NaN              NaN           NaN          NaN   NaN        NaN   
2           NaN              NaN           NaN          NaN   NaN        NaN   
3           NaN              NaN           NaN          NaN   NaN        NaN   
4           NaN              NaN           NaN          NaN   NaN        NaN   
...         ...              ...           ...          ...   ...        ...   
59187  softener              NaN           NaN  stretchable   NaN        NaN   
59188  softener              NaN           NaN  stretchable   NaN        NaN   
59189  softener              NaN           NaN  stretchable   NaN        NaN   
59190  softener              NaN           NaN  stretchable   NaN        NaN   
59191  softener              NaN           NaN  stretchable   NaN        NaN   

name    waist rise wash effect           wash quality work  
0              NaN         NaN                    NaN  NaN  
1              NaN         NaN                    NaN  NaN  
2              NaN         NaN                    NaN  NaN  
3              NaN         NaN                    NaN  NaN  
4              NaN         NaN                    NaN  NaN  
...            ...         ...                    ...  ...  
59187    high rise   non-faded               raw wash  NaN  
59188  medium rise   non-faded               bio wash  NaN  
59189  medium rise       faded  enzyme with acid wash  NaN  
59190  medium rise       faded  enzyme with acid wash  NaN  
59191  medium rise       faded  enzyme with acid wash  NaN  

[59192 rows x 98 columns]

In [7]:
# df_result = df_result.fillna("NaN")
# df_result

In [8]:
## Step for data transformation for final usage
key_columns = ['name','VariantId']
for col in df_result.columns:
    if col not in key_columns:
        df_result[col] = df_result.apply(lambda row: f"{{{col}: {row[col]}}}" if pd.notnull(row[col]) else "", axis=1)
df_result

name   VariantId Belt Type Buttons Closure Design               Fabric  \
0        1513399                                      {Fabric: Cotton}   
1        1530083                                   {Fabric: Polyester}   
2        1547548                                       {Fabric: Denim}   
3        1547567                                       {Fabric: Denim}   
4        1569064                                     {Fabric: Hosiery}   
...          ...       ...     ...     ...    ...                  ...   
59187    2652696                                                         
59188    2652713                                                         
59189    2652715                                                         
59190    2652716                                                         
59191    2652717                                                         

name  Fabric Composition             Fabric Grade             Fabric Weight  \
0                         {Fabric Grade: 200 GSM}                             
1                                                                             
2                                                                             
3                                                                             
4                                                  {Fabric Weight: 180 GSM}   
...                  ...                      ...                       ...   
59187                                                                         
59188                                                                         
59189                                                                         
59190                                                                         
59191                                                                         

name                  Fit  ...              softner stiching quality  \
0                          ...                                         
1                          ...                                         
2      {Fit: Regular Fit}  ...                                         
3      {Fit: Regular Fit}  ...                                         
4                          ...                                         
...                   ...  ...                  ...              ...   
59187                      ...  {softner: softener}                    
59188                      ...  {softner: softener}                    
59189                      ...  {softner: softener}                    
59190                      ...  {softner: softener}                    
59191                      ...  {softner: softener}                    

name  strechability                 stretch style top fabric  \
0                                                              
1                                                              
2                                                              
3                                                              
4                                                              
...             ...                     ...   ...        ...   
59187                {stretch: stretchable}                    
59188                {stretch: stretchable}                    
59189                {stretch: stretchable}                    
59190                {stretch: stretchable}                    
59191                {stretch: stretchable}                    

name                  waist rise               wash effect  \
0                                                            
1                                                            
2                                                            
3                                                            
4                                                            
...                          ...                       ...   
59187    {waist rise: high rise}  {wash effect: non-faded}   
59188  {waist rise: medium rise}

In [9]:
variant_ids = df_result[['VariantId']]
attributes = df_result.drop('VariantId', axis = 1)
encoded_attributes = pd.get_dummies(attributes)
encoded_attributes

Belt Type_  Belt Type_{Belt Type: Elastic}  Buttons_  \
0            True                           False      True   
1            True                           False      True   
2            True                           False      True   
3            True                           False      True   
4            True                           False      True   
...           ...                             ...       ...   
59187        True                           False      True   
59188        True                           False      True   
59189        True                           False      True   
59190        True                           False      True   
59191        True                           False      True   

       Buttons_{Buttons: 10 oz}  Buttons_{Buttons: 11 oz}  \
0                         False                     False   
1                         False                     False   
2                         False                     False   
3                         False                     False   
4                         False                     False   
...                         ...                       ...   
59187                     False                     False   
59188                     False                     False   
59189                     False                     False   
59190                     False                     False   
59191                     False                     False   

       Buttons_{Buttons: 2 Buttons}  Buttons_{Buttons: 3 Buttons}  \
0                             False                         False   
1                             False                         False   
2                             False                         False   
3                             False                         False   
4                             False                         False   
...                             ...                           ...   
59187                         False                         False   
59188                         False                         False   
59189                         False                         False   
59190                         False                         False   
59191                         False                         False   

       Buttons_{Buttons: 4 or More Buttons}  Buttons_{Buttons: 5 Buttons}  \
0                                     False                         False   
1                                     False                         False   
2                                     False                         False   
3                                     False                         False   
4                                     False                         False   
...                                     ...                           ...   
59187                                 False                         False   
59188                                 False                         False   
59189                                 False                         False   
59190                                 False                         False   
59191                                 False                         False   

       Buttons_{Buttons: N/A}  ...  wash quality_{wash quality: bio wash}  \
0                       False  ...                                  False   
1                       False  ...                                  False   
2                       False  ...                                  False   
3                       False  ...                                  False   
4                       False  ...                                  False   
...                       ...  ...                                    ...   
59187                   False  ...                                  False   
59188                   False  ...                                   True   
59189                   False  ...                                  Fa

In [10]:
encoded_attributes[encoded_attributes.select_dtypes(include=['bool']).columns] = encoded_attributes.select_dtypes(include=['bool']).astype(int)

In [11]:
encoded_attributes

Belt Type_  Belt Type_{Belt Type: Elastic}  Buttons_  \
0               1                               0         1   
1               1                               0         1   
2               1                               0         1   
3               1                               0         1   
4               1                               0         1   
...           ...                             ...       ...   
59187           1                               0         1   
59188           1                               0         1   
59189           1                               0         1   
59190           1                               0         1   
59191           1                               0         1   

       Buttons_{Buttons: 10 oz}  Buttons_{Buttons: 11 oz}  \
0                             0                         0   
1                             0                         0   
2                             0                         0   
3                             0                         0   
4                             0                         0   
...                         ...                       ...   
59187                         0                         0   
59188                         0                         0   
59189                         0                         0   
59190                         0                         0   
59191                         0                         0   

       Buttons_{Buttons: 2 Buttons}  Buttons_{Buttons: 3 Buttons}  \
0                                 0                             0   
1                                 0                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
...                             ...                           ...   
59187                             0                             0   
59188                             0                             0   
59189                             0                             0   
59190                             0                             0   
59191                             0                             0   

       Buttons_{Buttons: 4 or More Buttons}  Buttons_{Buttons: 5 Buttons}  \
0                                         0                             0   
1                                         0                             0   
2                                         0                             0   
3                                         0                             0   
4                                         0                             0   
...                                     ...                           ...   
59187                                     0                             0   
59188                                     0                             0   
59189                                     0                             0   
59190                                     0                             0   
59191                                     0                             0   

       Buttons_{Buttons: N/A}  ...  wash quality_{wash quality: bio wash}  \
0                           0  ...                                      0   
1                           0  ...                                      0   
2                           0  ...                                      0   
3                           0  ...                                      0   
4                           0  ...                                      0   
...                       ...  ...                                    ...   
59187                       0  ...                                      0   
59188                       0  ...                                      1   
59189                       0  ...                                    

In [12]:
df_result = pd.concat([variant_ids, encoded_attributes], axis=1)
df_result

VariantId  Belt Type_  Belt Type_{Belt Type: Elastic}  Buttons_  \
0        1513399           1                               0         1   
1        1530083           1                               0         1   
2        1547548           1                               0         1   
3        1547567           1                               0         1   
4        1569064           1                               0         1   
...          ...         ...                             ...       ...   
59187    2652696           1                               0         1   
59188    2652713           1                               0         1   
59189    2652715           1                               0         1   
59190    2652716           1                               0         1   
59191    2652717           1                               0         1   

       Buttons_{Buttons: 10 oz}  Buttons_{Buttons: 11 oz}  \
0                             0                         0   
1                             0                         0   
2                             0                         0   
3                             0                         0   
4                             0                         0   
...                         ...                       ...   
59187                         0                         0   
59188                         0                         0   
59189                         0                         0   
59190                         0                         0   
59191                         0                         0   

       Buttons_{Buttons: 2 Buttons}  Buttons_{Buttons: 3 Buttons}  \
0                                 0                             0   
1                                 0                             0   
2                                 0                             0   
3                                 0                             0   
4                                 0                             0   
...                             ...                           ...   
59187                             0                             0   
59188                             0                             0   
59189                             0                             0   
59190                             0                             0   
59191                             0                             0   

       Buttons_{Buttons: 4 or More Buttons}  Buttons_{Buttons: 5 Buttons}  \
0                                         0                             0   
1                                         0                             0   
2                                         0                             0   
3                                         0                             0   
4                                         0                             0   
...                                     ...                           ...   
59187                                     0                             0   
59188                                     0                             0   
59189                                     0                             0   
59190                                     0                             0   
59191                                     0                             0   

       ...  wash quality_{wash quality: bio wash}  \
0      ...                                      0   
1      ...                                      0   
2      ...                                      0   
3      ...                                      0   
4      ...                                      0   
...    ...                                    ...   
59187  ...                                      0   
59188  ...                                      1   
59189  ...                                      0   
59190  ...                                      0   
59191  ...                                      

In [13]:
df_final_new = pd.merge(df,df_result,on='VariantId',how='left')
df_final_new

VariantId                                   Color ColorHex  Category  \
0        1513399                                   CREAM  #FFFFCC  t-shirts   
1        1530083  BLACK-1, GOLDEN-1, RED-1, ROYAL BLUE-1           t-shirts   
2        1547548                                   KHAKI  #a18f5e     jeans   
3        1547567                                    BLUE  #0083C3     jeans   
4        1569064                            MULTICOLOR-5           t-shirts   
...          ...                                     ...      ...       ...   
59187    2652696                                SKY BLUE  #74D1EA     jeans   
59188    2652713                                    GREY  #adadad     jeans   
59189    2652715                                SKY BLUE  #74D1EA     jeans   
59190    2652716                                    BLUE  #0083C3     jeans   
59191    2652717                               DARK GREY  #545454     jeans   

      MainCategory SuperCategory  \
0              Men      Apparels   
1              Men      Apparels   
2      Boys - Kids      Apparels   
3      Boys - Kids      Apparels   
4      Boys - Kids      Apparels   
...            ...           ...   
59187        Women      Apparels   
59188  Boys - Kids      Apparels   
59189          Men      Apparels   
59190          Men      Apparels   
59191          Men      Apparels   

                                        ProductAttribute  Belt Type_  \
0      [{'name': 'Fabric', 'value': 'Cotton', 'valueI...           1   
1      [{'name': 'Fabric', 'value': 'Polyester', 'val...           1   
2      [{'name': 'Style', 'value': 'Embroidered Jeans...           1   
3      [{'name': 'Style', 'value': 'Embroidered Jeans...           1   
4      [{'name': 'Style', 'value': 'Solid / Plain T-S...           1   
...                                                  ...         ...   
59187  [{'name': 'fabric', 'value': 'Cotton Poly 2/1 ...           1   
59188  [{'name': 'fabric', 'value': 'Cotton Lycra Kni...           1   
59189  [{'name': 'wash quality', 'value': 'enzyme wit...           1   
59190  [{'name': 'wash quality', 'value': 'enzyme wit...           1   
59191  [{'name': 'wash quality', 'value': 'enzyme wit...           1   

       Belt Type_{Belt Type: Elastic}  Buttons_  ...  \
0                                   0         1  ...   
1                                   0         1  ...   
2                                   0         1  ...   
3                                   0         1  ...   
4                                   0         1  ...   
...                               ...       ...  ...   
59187                               0         1  ...   
59188                               0         1  ...   
59189                               0         1  ...   
59190                               0         1  ...   
59191                               0         1  ...   

       wash quality_{wash quality: bio wash}  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   
...                                      ...   
59187                                      0   
59188                                      1   
59189                                      0   
59190                                      0   
59191                                      0   

       wash quality_{wash quality: enzyme wash}  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   
...                                         ...   
59187                                         0   
59188                                         0   
59189                                         0   


In [14]:
variants_list = df_final_new['VariantId'].to_list()

In [15]:
import pandas as pd
from pymongo import MongoClient
import concurrent.futures
from tqdm import tqdm

# MongoDB connection setup
client = MongoClient("mongodb+srv://BijnisLiveMongoReader:XUerT9JxnWnEstSA@prod-mongo.bijnis.com/bijnis_production?tls=false&ssl=false")  # Replace with your connection string
db = client["product_service_production"]  # Replace with your database name
col = db["variant_image_vectors"]  # Replace with your collection name

# Define the query filter
query_filter = {"variantId": {'$in': variants_list}}

# Function to fetch documents in batches
def fetch_documents(skip, limit):
    return list(col.find(query_filter).skip(skip).limit(limit))

# Function to fetch all documents using threading with progress bar
def fetch_all_documents(total_records, batch_size):
    all_documents = []
    num_batches = total_records // batch_size + (1 if total_records % batch_size > 0 else 0)  # Calculate total batches
    
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = {executor.submit(fetch_documents, skip, batch_size): skip for skip in range(0, total_records, batch_size)}
        
        # Use tqdm to show progress bar
        for future in tqdm(concurrent.futures.as_completed(futures), total=num_batches, desc="Fetching documents"):
            try:
                result = future.result()
                all_documents.extend(result)
            except Exception as e:
                print(f"Error fetching documents: {e}")
                
    return all_documents

# Fetching the total count of documents that match the query
total_records = col.count_documents(query_filter)

# Define the batch size
batch_size = 10000  # Adjust this according to your memory capacity

# Fetch all documents
all_docs = fetch_all_documents(total_records, batch_size)

# Convert the result to a DataFrame
df_mongo = pd.DataFrame(all_docs)

# Display the DataFrame
df_mongo 

Fetching documents: 100%|██████████| 25/25 [14:48<00:00, 35.54s/it]  


_id  variantId   imageId  \
0       66e7803df37ca712f0413a2d    2307410  22147004   
1       66e7803df37ca712f0413a2e    2307410  22147005   
2       66e7803df37ca712f0413a2f    2307410  22147006   
3       66e7803df37ca712f0413a30    2307410  22147007   
4       66e7803df37ca712f0413a31    2307410  22147008   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
1       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
2       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
3       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
4       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector            createdOn  
0       [0.02699310891330242, 0.026698676869273186, 0....  2024-09-16 06:17:57  
1       [0.15216664969921112, 0.7531224489212036, 0.78...  2024-09-16 06:17:57  
2       [0.27668434381484985, 0.0, 0.6610189080238342,...  2024-09-16 06:17:57  
3       [0.0, 0.33198481798171997, 0.0, 0.0, 0.1082259...  2024-09-16 06:17:57  
4       [0.015320106409490108, 0.13437797129154205, 0....  2024-09-16 06:17:57  
...                                                   ...                  ...  
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...  2024-09-30 09:14:51  
247119  [0.12077472358942032, 0.0, 0.04706697165966034...  2024-09-30 09:14:41  
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...  2024-09-30 09:14:41  
247121  [0.015275663696229458, 0.0, 0.0472753793001174...  2024-09-30 09:14:36  
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...  2024-09-30 09:14:36  

[247123 rows x 6 columns]

In [16]:
df_mongo = df_mongo.rename(columns={"variantId":"VariantId"})
df_mongo

_id  VariantId   imageId  \
0       66e7803df37ca712f0413a2d    2307410  22147004   
1       66e7803df37ca712f0413a2e    2307410  22147005   
2       66e7803df37ca712f0413a2f    2307410  22147006   
3       66e7803df37ca712f0413a30    2307410  22147007   
4       66e7803df37ca712f0413a31    2307410  22147008   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
1       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
2       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
3       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
4       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector            createdOn  
0       [0.02699310891330242, 0.026698676869273186, 0....  2024-09-16 06:17:57  
1       [0.15216664969921112, 0.7531224489212036, 0.78...  2024-09-16 06:17:57  
2       [0.27668434381484985, 0.0, 0.6610189080238342,...  2024-09-16 06:17:57  
3       [0.0, 0.33198481798171997, 0.0, 0.0, 0.1082259...  2024-09-16 06:17:57  
4       [0.015320106409490108, 0.13437797129154205, 0....  2024-09-16 06:17:57  
...                                                   ...                  ...  
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...  2024-09-30 09:14:51  
247119  [0.12077472358942032, 0.0, 0.04706697165966034...  2024-09-30 09:14:41  
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...  2024-09-30 09:14:41  
247121  [0.015275663696229458, 0.0, 0.0472753793001174...  2024-09-30 09:14:36  
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...  2024-09-30 09:14:36  

[247123 rows x 6 columns]

In [17]:
df_flattened = pd.merge(df_mongo, df_final_new, on = 'VariantId', how='left')

In [18]:
df_flattened

_id  VariantId   imageId  \
0       66e7803df37ca712f0413a2d    2307410  22147004   
1       66e7803df37ca712f0413a2e    2307410  22147005   
2       66e7803df37ca712f0413a2f    2307410  22147006   
3       66e7803df37ca712f0413a30    2307410  22147007   
4       66e7803df37ca712f0413a31    2307410  22147008   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
1       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
2       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
3       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
4       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector  \
0       [0.02699310891330242, 0.026698676869273186, 0....   
1       [0.15216664969921112, 0.7531224489212036, 0.78...   
2       [0.27668434381484985, 0.0, 0.6610189080238342,...   
3       [0.0, 0.33198481798171997, 0.0, 0.0, 0.1082259...   
4       [0.015320106409490108, 0.13437797129154205, 0....   
...                                                   ...   
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...   
247119  [0.12077472358942032, 0.0, 0.04706697165966034...   
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...   
247121  [0.015275663696229458, 0.0, 0.0472753793001174...   
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...   

                  createdOn         Color ColorHex  Category MainCategory  \
0       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
1       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
2       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
3       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
4       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
...                     ...           ...      ...       ...          ...   
247118  2024-09-30 09:14:51    MULTICOLOR  #999999     jeans          Men   
247119  2024-09-30 09:14:41    MULTICOLOR  #999999     jeans          Men   
247120  2024-09-30 09:14:41    MULTICOLOR  #999999     jeans          Men   
247121  2024-09-30 09:14:36    MULTICOLOR  #999999     jeans          Men   
247122  2024-09-30 09:14:36    MULTICOLOR  #999999     jeans          Men   

        ... wash quality_{wash quality: bio wash}  \
0       ...                                     0   
1       ...                                     0   
2       ...                                     0   
3       ...                                     0   
4       ...                                     0   
...     ...                                   ...   
247118  ...                                     0   
247119  ...                                     0   
247120  ...                                     0   
247121  ...                                     0   
247122  ...                                     0   

       wash quality_{wash quality: enzyme wash}  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   
...       

In [19]:
## Transformation of the vector column to individual labels for training purpose
vector_df = pd.DataFrame(df_flattened['vector'].to_list())
vector_df.columns = [f"V{i+1}" for i in range (vector_df.shape[1])]
df_expanded = pd.concat([df_flattened,vector_df],axis=1)
df_expanded = df_expanded.fillna(0)
df_expanded

_id  VariantId   imageId  \
0       66e7803df37ca712f0413a2d    2307410  22147004   
1       66e7803df37ca712f0413a2e    2307410  22147005   
2       66e7803df37ca712f0413a2f    2307410  22147006   
3       66e7803df37ca712f0413a30    2307410  22147007   
4       66e7803df37ca712f0413a31    2307410  22147008   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
1       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
2       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
3       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
4       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector  \
0       [0.02699310891330242, 0.026698676869273186, 0....   
1       [0.15216664969921112, 0.7531224489212036, 0.78...   
2       [0.27668434381484985, 0.0, 0.6610189080238342,...   
3       [0.0, 0.33198481798171997, 0.0, 0.0, 0.1082259...   
4       [0.015320106409490108, 0.13437797129154205, 0....   
...                                                   ...   
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...   
247119  [0.12077472358942032, 0.0, 0.04706697165966034...   
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...   
247121  [0.015275663696229458, 0.0, 0.0472753793001174...   
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...   

                  createdOn         Color ColorHex  Category MainCategory  \
0       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
1       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
2       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
3       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
4       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
...                     ...           ...      ...       ...          ...   
247118  2024-09-30 09:14:51    MULTICOLOR  #999999     jeans          Men   
247119  2024-09-30 09:14:41    MULTICOLOR  #999999     jeans          Men   
247120  2024-09-30 09:14:41    MULTICOLOR  #999999     jeans          Men   
247121  2024-09-30 09:14:36    MULTICOLOR  #999999     jeans          Men   
247122  2024-09-30 09:14:36    MULTICOLOR  #999999     jeans          Men   

        ...     V1271     V1272     V1273     V1274     V1275     V1276  \
0       ...  0.000000  0.502343  0.000000  0.099487  0.002177  0.436013   
1       ...  0.000000  0.839816  0.434324  0.000000  0.405357  0.068908   
2       ...  0.000000  0.000000  0.003457  0.100403  0.544899  0.000000   
3       ...  0.000000  1.259930  0.078931  0.005840  0.000000  0.011345   
4       ...  0.000000  1.100698  0.072530  0.029821  0.139721  0.020297   
...     ...       ...       ...       ...       ...       ...       ...   
247118  ...  0.015857  0.292787  0.000000  0.636535  0.230335  0.000000   
247119  ...  0.000000  1.123815  0.000000  0.097913  0.000000  0.000000   
247120  ...  0.000000  0.885310  0.000000  0.239653  0.439827  0.004465   
247121  ...  0.000000  0.314013  0.000000  0.023167  0.016230  0.000000   
247122  ...  0.000000  0.269158  0.000000  0.014829  0.187695  0.000000   

           V1277     V1278     V1279     V1280  
0  

In [20]:
#df_expanded.to_csv('Classification_Stage_1.csv')

In [21]:
## creating version_1 of the dataframe for further operations
filtered_columns = [col for col in df_expanded.columns if col.startswith('V') and col[1:].isdigit()] + ['Category']
filtered_df = df_expanded[filtered_columns]
filtered_df

V1        V2        V3        V4        V5        V6        V7  \
0       0.026993  0.026699  0.013406  0.158621  0.000000  0.047574  1.477892   
1       0.152167  0.753122  0.780756  0.033834  0.224728  0.538430  0.347718   
2       0.276684  0.000000  0.661019  0.000000  0.000000  0.006237  0.270860   
3       0.000000  0.331985  0.000000  0.000000  0.108226  0.138597  1.015925   
4       0.015320  0.134378  0.000000  0.000000  0.001563  0.295301  0.045144   
...          ...       ...       ...       ...       ...       ...       ...   
247118  0.000000  0.000000  0.119957  0.197625  0.000000  0.000000  0.213593   
247119  0.120775  0.000000  0.047067  0.145077  0.000000  0.294627  0.264174   
247120  0.000000  0.000000  0.530862  0.009705  0.000000  0.000000  0.130015   
247121  0.015276  0.000000  0.047275  0.000000  0.020218  0.047031  0.177078   
247122  0.000000  0.000000  0.015624  0.012814  0.000000  0.000000  0.331728   

              V8        V9       V10  ...     V1272     V1273     V1274  \
0       0.136192  0.215258  0.034348  ...  0.502343  0.000000  0.099487   
1       0.000000  0.061816  0.000000  ...  0.839816  0.434324  0.000000   
2       0.095041  0.000000  0.736817  ...  0.000000  0.003457  0.100403   
3       0.000000  0.013612  0.108900  ...  1.259930  0.078931  0.005840   
4       0.176906  0.028280  0.000000  ...  1.100698  0.072530  0.029821   
...          ...       ...       ...  ...       ...       ...       ...   
247118  0.148386  0.214779  0.542384  ...  0.292787  0.000000  0.636535   
247119  0.000000  1.250417  0.036951  ...  1.123815  0.000000  0.097913   
247120  0.477553  0.000000  0.833972  ...  0.885310  0.000000  0.239653   
247121  0.000000  0.345184  0.066882  ...  0.314013  0.000000  0.023167   
247122  0.661369  0.129900  0.172631  ...  0.269158  0.000000  0.014829   

           V1275     V1276     V1277     V1278     V1279     V1280  Category  
0       0.002177  0.436013  0.378547  0.029962  0.127837  0.000000  t-shirts  
1       0.405357  0.068908  0.269205  0.000000  0.325799  0.036937  t-shirts  
2       0.544899  0.000000  0.000000  0.267199  0.880283  0.492699  t-shirts  
3       0.000000  0.011345  0.000000  0.000000  0.478731  0.518241  t-shirts  
4       0.139721  0.020297  0.000000  0.000000  0.019781  0.000000  t-shirts  
...          ...       ...       ...       ...       ...       ...       ...  
247118  0.230335  0.000000  0.289987  1.239995  0.132540  0.270762     jeans  
247119  0.000000  0.000000  0.023646  0.102291  0.582427  0.024401     jeans  
247120  0.439827  0.004465  0.000000  0.048496  0.211974  0.088579     jeans  
247121  0.016230  0.000000  0.000000  0.000000  0.000000  0.147468     jeans  
247122  0.187695  0.000000  0.002154  0.486474  0.748830  0.622428     jeans  

[247123 rows x 1281 columns]

In [22]:
filtered_df['Category'] = filtered_df['Category'].str.upper()
# filtered_df['MainCategory'] = filtered_df['MainCategory'].str.upper()
# filtered_df['SuperCategory'] = filtered_df['SuperCategory'].str.upper()
filtered_df

C:\Users\Admin\AppData\Local\Temp\ipykernel_22140\3972431576.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df['Category'] = filtered_df['Category'].str.upper()


V1        V2        V3        V4        V5        V6        V7  \
0       0.026993  0.026699  0.013406  0.158621  0.000000  0.047574  1.477892   
1       0.152167  0.753122  0.780756  0.033834  0.224728  0.538430  0.347718   
2       0.276684  0.000000  0.661019  0.000000  0.000000  0.006237  0.270860   
3       0.000000  0.331985  0.000000  0.000000  0.108226  0.138597  1.015925   
4       0.015320  0.134378  0.000000  0.000000  0.001563  0.295301  0.045144   
...          ...       ...       ...       ...       ...       ...       ...   
247118  0.000000  0.000000  0.119957  0.197625  0.000000  0.000000  0.213593   
247119  0.120775  0.000000  0.047067  0.145077  0.000000  0.294627  0.264174   
247120  0.000000  0.000000  0.530862  0.009705  0.000000  0.000000  0.130015   
247121  0.015276  0.000000  0.047275  0.000000  0.020218  0.047031  0.177078   
247122  0.000000  0.000000  0.015624  0.012814  0.000000  0.000000  0.331728   

              V8        V9       V10  ...     V1272     V1273     V1274  \
0       0.136192  0.215258  0.034348  ...  0.502343  0.000000  0.099487   
1       0.000000  0.061816  0.000000  ...  0.839816  0.434324  0.000000   
2       0.095041  0.000000  0.736817  ...  0.000000  0.003457  0.100403   
3       0.000000  0.013612  0.108900  ...  1.259930  0.078931  0.005840   
4       0.176906  0.028280  0.000000  ...  1.100698  0.072530  0.029821   
...          ...       ...       ...  ...       ...       ...       ...   
247118  0.148386  0.214779  0.542384  ...  0.292787  0.000000  0.636535   
247119  0.000000  1.250417  0.036951  ...  1.123815  0.000000  0.097913   
247120  0.477553  0.000000  0.833972  ...  0.885310  0.000000  0.239653   
247121  0.000000  0.345184  0.066882  ...  0.314013  0.000000  0.023167   
247122  0.661369  0.129900  0.172631  ...  0.269158  0.000000  0.014829   

           V1275     V1276     V1277     V1278     V1279     V1280  Category  
0       0.002177  0.436013  0.378547  0.029962  0.127837  0.000000  T-SHIRTS  
1       0.405357  0.068908  0.269205  0.000000  0.325799  0.036937  T-SHIRTS  
2       0.544899  0.000000  0.000000  0.267199  0.880283  0.492699  T-SHIRTS  
3       0.000000  0.011345  0.000000  0.000000  0.478731  0.518241  T-SHIRTS  
4       0.139721  0.020297  0.000000  0.000000  0.019781  0.000000  T-SHIRTS  
...          ...       ...       ...       ...       ...       ...       ...  
247118  0.230335  0.000000  0.289987  1.239995  0.132540  0.270762     JEANS  
247119  0.000000  0.000000  0.023646  0.102291  0.582427  0.024401     JEANS  
247120  0.439827  0.004465  0.000000  0.048496  0.211974  0.088579     JEANS  
247121  0.016230  0.000000  0.000000  0.000000  0.000000  0.147468     JEANS  
247122  0.187695  0.000000  0.002154  0.486474  0.748830  0.622428     JEANS  

[247123 rows x 1281 columns]

### Start of the classification code
    > First Attempt with RandomForestClassifier

In [23]:
filtered_df

V1        V2        V3        V4        V5        V6        V7  \
0       0.026993  0.026699  0.013406  0.158621  0.000000  0.047574  1.477892   
1       0.152167  0.753122  0.780756  0.033834  0.224728  0.538430  0.347718   
2       0.276684  0.000000  0.661019  0.000000  0.000000  0.006237  0.270860   
3       0.000000  0.331985  0.000000  0.000000  0.108226  0.138597  1.015925   
4       0.015320  0.134378  0.000000  0.000000  0.001563  0.295301  0.045144   
...          ...       ...       ...       ...       ...       ...       ...   
247118  0.000000  0.000000  0.119957  0.197625  0.000000  0.000000  0.213593   
247119  0.120775  0.000000  0.047067  0.145077  0.000000  0.294627  0.264174   
247120  0.000000  0.000000  0.530862  0.009705  0.000000  0.000000  0.130015   
247121  0.015276  0.000000  0.047275  0.000000  0.020218  0.047031  0.177078   
247122  0.000000  0.000000  0.015624  0.012814  0.000000  0.000000  0.331728   

              V8        V9       V10  ...     V1272     V1273     V1274  \
0       0.136192  0.215258  0.034348  ...  0.502343  0.000000  0.099487   
1       0.000000  0.061816  0.000000  ...  0.839816  0.434324  0.000000   
2       0.095041  0.000000  0.736817  ...  0.000000  0.003457  0.100403   
3       0.000000  0.013612  0.108900  ...  1.259930  0.078931  0.005840   
4       0.176906  0.028280  0.000000  ...  1.100698  0.072530  0.029821   
...          ...       ...       ...  ...       ...       ...       ...   
247118  0.148386  0.214779  0.542384  ...  0.292787  0.000000  0.636535   
247119  0.000000  1.250417  0.036951  ...  1.123815  0.000000  0.097913   
247120  0.477553  0.000000  0.833972  ...  0.885310  0.000000  0.239653   
247121  0.000000  0.345184  0.066882  ...  0.314013  0.000000  0.023167   
247122  0.661369  0.129900  0.172631  ...  0.269158  0.000000  0.014829   

           V1275     V1276     V1277     V1278     V1279     V1280  Category  
0       0.002177  0.436013  0.378547  0.029962  0.127837  0.000000  T-SHIRTS  
1       0.405357  0.068908  0.269205  0.000000  0.325799  0.036937  T-SHIRTS  
2       0.544899  0.000000  0.000000  0.267199  0.880283  0.492699  T-SHIRTS  
3       0.000000  0.011345  0.000000  0.000000  0.478731  0.518241  T-SHIRTS  
4       0.139721  0.020297  0.000000  0.000000  0.019781  0.000000  T-SHIRTS  
...          ...       ...       ...       ...       ...       ...       ...  
247118  0.230335  0.000000  0.289987  1.239995  0.132540  0.270762     JEANS  
247119  0.000000  0.000000  0.023646  0.102291  0.582427  0.024401     JEANS  
247120  0.439827  0.004465  0.000000  0.048496  0.211974  0.088579     JEANS  
247121  0.016230  0.000000  0.000000  0.000000  0.000000  0.147468     JEANS  
247122  0.187695  0.000000  0.002154  0.486474  0.748830  0.622428     JEANS  

[247123 rows x 1281 columns]

In [24]:
from sklearn.model_selection import train_test_split 
X = filtered_df.drop(['Category'], axis = 1)
y = filtered_df[['Category']]
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state=42)

In [25]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)



c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


RandomForestClassifier(random_state=42)

In [26]:
from sklearn.metrics import accuracy_score, f1_score
# Predict on test data
y_pred = rf.predict(X_test)

# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred, average='weighted'))

Accuracy: 0.9869499241274658
F1 Score: 0.9869373625981873


In [27]:

import os
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.preprocessing.image import img_to_array
from PIL import Image, UnidentifiedImageError
import requests
from io import BytesIO
import streamlit as st
import json
import datetime


def extract_feature_vector(image_path):
    # Load the pre-trained MobileNetV2 model
    base_model = MobileNetV2(weights='imagenet', include_top=False, pooling='avg')
    model = tf.keras.Model(inputs=base_model.input, outputs=base_model.output)

    # Open the image using PIL
    img = Image.open(image_path)

    if img.format != 'JPEG':
        # Convert image to RGB if it is not
        if img.mode != 'RGB':
            img = img.convert('RGB')
        
        img.save(image_path, format='JPEG')

    # Resize the image to the size expected by MobileNetV2 (224x224)
    img = img.resize((224, 224))
    # Convert the image to a numpy array and preprocess for the model
    img_array = np.array(img)
    img_array = tf.keras.applications.mobilenet_v2.preprocess_input(img_array)
    img_array = np.expand_dims(img_array, axis=0)  # Add batch dimension

    # Extract the feature vector using the model
    feature_vector = model.predict(img_array)

    return feature_vector

# Example usage
image_path = r"C:\Users\Admin\Attribute_Classification\Product_Attribute_Classification\Image_Vectorization\Image_10.jpeg"
feature_vector = extract_feature_vector(image_path)
image_vector_df = pd.DataFrame(feature_vector)
image_vector_df.columns = [f"V{i+1}" for i in range (image_vector_df.shape[1])]
image_vector_df




1/1 [==============================] - 2s 2s/step


V1   V2        V3       V4   V5   V6        V7        V8   V9    V10  ...  \
0  0.0  0.0  0.094462  0.08731  0.0  0.0  0.003508  0.047143  0.0  0.003  ...   

   V1271     V1272  V1273  V1274     V1275     V1276  V1277     V1278  V1279  \
0    0.0  0.564352    0.0    0.0  0.462524  0.023598    0.0  0.496213    0.0   

   V1280  
0    0.0  

[1 rows x 1280 columns]

In [28]:
y_new = rf.predict(image_vector_df)
y_new

array(['JEANS'], dtype=object)

In [29]:
import pickle

In [30]:
with open('random_forest_model.pkl', 'wb') as f:
    pickle.dump(rf, f)

In [31]:
with open('random_forest_model.pkl', 'rb') as f:
    rf = pickle.load(f)

In [32]:

import os
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.preprocessing.image import img_to_array
from PIL import Image, UnidentifiedImageError
import requests
from io import BytesIO
import streamlit as st
import json
import datetime


def extract_feature_vector(image_path):
    # Load the pre-trained MobileNetV2 model
    base_model = MobileNetV2(weights='imagenet', include_top=False, pooling='avg')
    model = tf.keras.Model(inputs=base_model.input, outputs=base_model.output)

    # Open the image using PIL
    img = Image.open(image_path)

    if img.format != 'JPEG':
        # Convert image to RGB if it is not
        if img.mode != 'RGB':
            img = img.convert('RGB')
        
        img.save(image_path, format='JPEG')

    # Resize the image to the size expected by MobileNetV2 (224x224)
    img = img.resize((224, 224))
    # Convert the image to a numpy array and preprocess for the model
    img_array = np.array(img)
    img_array = tf.keras.applications.mobilenet_v2.preprocess_input(img_array)
    img_array = np.expand_dims(img_array, axis=0)  # Add batch dimension

    # Extract the feature vector using the model
    feature_vector = model.predict(img_array)

    return feature_vector

# Example usage
image_path = r"C:\Users\Admin\Attribute_Classification\Product_Attribute_Classification\Image_Vectorization\Image_11.jpeg"
feature_vector = extract_feature_vector(image_path)
image_vector_df = pd.DataFrame(feature_vector)
image_vector_df.columns = [f"V{i+1}" for i in range (image_vector_df.shape[1])]
image_vector_df




1/1 [==============================] - 1s 1s/step


V1   V2   V3      V4   V5        V6        V7        V8        V9  V10  \
0  0.0  0.0  0.0  2.3025  0.0  0.030358  0.088443  0.034139  1.260753  0.0   

   ...  V1271     V1272  V1273     V1274  V1275  V1276     V1277     V1278  \
0  ...    0.0  0.114412    0.0  0.079145    0.0    0.0  0.678712  0.351788   

   V1279     V1280  
0    0.0  0.014979  

[1 rows x 1280 columns]

In [33]:
y_new_2 = rf.predict(image_vector_df)
y_new_2

array(['JEANS'], dtype=object)

    > Trying SGDClassifier (Stochastic Gradient Descent) with incremental learning using chunks for faster and efficient execution

In [34]:
from sklearn.linear_model import SGDClassifier
sgd_clf = SGDClassifier(max_iter=1000, tol=1e-3, random_state=42)

# Scaling in chunks 
batch_size = 10000
for start in range(0, len(X_train),batch_size):
    end = start + batch_size
    X_batch = X_train[start:end]
    y_batch = y_train[start:end]
    
    sgd_clf.partial_fit(X_batch,y_batch, classes = np.unique(y))

y_pred_1 = sgd_clf.predict(X_test)

c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was e

In [35]:
# Accuracy Testing 
from sklearn.metrics import accuracy_score, f1_score
# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_1))
print("F1 Score:", f1_score(y_test, y_pred_1, average='weighted'))

Accuracy: 0.9805159332321699
F1 Score: 0.9804526107895285


    > SGDClassifier has significantly reduced the training time with marginal change (~ 0.50 %) in the accuracy

### Code for the Color Classification

    > Trying first time with the SGDClassifier for color classification and training

In [38]:
df_expanded

_id  VariantId   imageId  \
0       66e7803df37ca712f0413a2d    2307410  22147004   
1       66e7803df37ca712f0413a2e    2307410  22147005   
2       66e7803df37ca712f0413a2f    2307410  22147006   
3       66e7803df37ca712f0413a30    2307410  22147007   
4       66e7803df37ca712f0413a31    2307410  22147008   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
1       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
2       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
3       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
4       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector  \
0       [0.02699310891330242, 0.026698676869273186, 0....   
1       [0.15216664969921112, 0.7531224489212036, 0.78...   
2       [0.27668434381484985, 0.0, 0.6610189080238342,...   
3       [0.0, 0.33198481798171997, 0.0, 0.0, 0.1082259...   
4       [0.015320106409490108, 0.13437797129154205, 0....   
...                                                   ...   
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...   
247119  [0.12077472358942032, 0.0, 0.04706697165966034...   
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...   
247121  [0.015275663696229458, 0.0, 0.0472753793001174...   
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...   

                  createdOn         Color ColorHex  Category MainCategory  \
0       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
1       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
2       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
3       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
4       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
...                     ...           ...      ...       ...          ...   
247118  2024-09-30 09:14:51    MULTICOLOR  #999999     jeans          Men   
247119  2024-09-30 09:14:41    MULTICOLOR  #999999     jeans          Men   
247120  2024-09-30 09:14:41    MULTICOLOR  #999999     jeans          Men   
247121  2024-09-30 09:14:36    MULTICOLOR  #999999     jeans          Men   
247122  2024-09-30 09:14:36    MULTICOLOR  #999999     jeans          Men   

        ...     V1271     V1272     V1273     V1274     V1275     V1276  \
0       ...  0.000000  0.502343  0.000000  0.099487  0.002177  0.436013   
1       ...  0.000000  0.839816  0.434324  0.000000  0.405357  0.068908   
2       ...  0.000000  0.000000  0.003457  0.100403  0.544899  0.000000   
3       ...  0.000000  1.259930  0.078931  0.005840  0.000000  0.011345   
4       ...  0.000000  1.100698  0.072530  0.029821  0.139721  0.020297   
...     ...       ...       ...       ...       ...       ...       ...   
247118  ...  0.015857  0.292787  0.000000  0.636535  0.230335  0.000000   
247119  ...  0.000000  1.123815  0.000000  0.097913  0.000000  0.000000   
247120  ...  0.000000  0.885310  0.000000  0.239653  0.439827  0.004465   
247121  ...  0.000000  0.314013  0.000000  0.023167  0.016230  0.000000   
247122  ...  0.000000  0.269158  0.000000  0.014829  0.187695  0.000000   

           V1277     V1278     V1279     V1280  
0  

In [37]:
import mysql.connector
import pandas as pd
mydb=mysql.connector.connect(host='report-reader.cluster-custom-c8oe0gvszktr.ap-south-1.rds.amazonaws.com',
                             user='skreadonlyuser_api',
                             password='XUerT9JxnWnEstSA')
cursor=mydb.cursor()

In [77]:
# cursor.execute('''
# Select 
# v.Variantid as VariantId,
# clr.ColorName as ColorName,
# clr.colorHex as ColorHex
# from shoekonnect_live.variants  as v
# INNER JOIN shoekonnect_live.colors as clr on clr.uniqueId = v.colorId
# ''')

In [78]:
# import pandas as pd
# import json
# rows=cursor.fetchall()
# columns=[i[0] for i in  cursor.description]
# df_colors=pd.DataFrame(rows,columns=columns)
# df_colors

VariantId   ColorName ColorHex
0          1000000      Coffee  #4a342e
1          1000001       Black  #333333
2          1000002      Coffee  #4a342e
3          1000003     Mustard  #e1b228
4          1000004      Coffee  #4a342e
...            ...         ...      ...
1632912    2652541  Multicolor  #999999
1632913    2652542  Multicolor  #999999
1632914    2652543  Multicolor  #999999
1632915    2652544        Pink  #ec81b3
1632916    2652545  Multicolor  #999999

[1632917 rows x 3 columns]

In [79]:
# df_colors['ColorName'] = df_colors['ColorName'].str.upper()
# df_colors

VariantId   ColorName ColorHex
0          1000000      COFFEE  #4a342e
1          1000001       BLACK  #333333
2          1000002      COFFEE  #4a342e
3          1000003     MUSTARD  #e1b228
4          1000004      COFFEE  #4a342e
...            ...         ...      ...
1632912    2652541  MULTICOLOR  #999999
1632913    2652542  MULTICOLOR  #999999
1632914    2652543  MULTICOLOR  #999999
1632915    2652544        PINK  #ec81b3
1632916    2652545  MULTICOLOR  #999999

[1632917 rows x 3 columns]

In [39]:
# color_class = pd.merge(df_expanded,df_colors, on='VariantId',how='left')
# color_class

NameError: name 'df_colors' is not defined

In [43]:
filtered_columns = [col for col in df_expanded.columns if col.startswith('V') and col[1:].isdigit()] + ['Color']
filtered_color_class = df_expanded[filtered_columns]
filtered_color_class

V1        V2        V3        V4        V5        V6        V7  \
0       0.026993  0.026699  0.013406  0.158621  0.000000  0.047574  1.477892   
1       0.152167  0.753122  0.780756  0.033834  0.224728  0.538430  0.347718   
2       0.276684  0.000000  0.661019  0.000000  0.000000  0.006237  0.270860   
3       0.000000  0.331985  0.000000  0.000000  0.108226  0.138597  1.015925   
4       0.015320  0.134378  0.000000  0.000000  0.001563  0.295301  0.045144   
...          ...       ...       ...       ...       ...       ...       ...   
247118  0.000000  0.000000  0.119957  0.197625  0.000000  0.000000  0.213593   
247119  0.120775  0.000000  0.047067  0.145077  0.000000  0.294627  0.264174   
247120  0.000000  0.000000  0.530862  0.009705  0.000000  0.000000  0.130015   
247121  0.015276  0.000000  0.047275  0.000000  0.020218  0.047031  0.177078   
247122  0.000000  0.000000  0.015624  0.012814  0.000000  0.000000  0.331728   

              V8        V9       V10  ...     V1272     V1273     V1274  \
0       0.136192  0.215258  0.034348  ...  0.502343  0.000000  0.099487   
1       0.000000  0.061816  0.000000  ...  0.839816  0.434324  0.000000   
2       0.095041  0.000000  0.736817  ...  0.000000  0.003457  0.100403   
3       0.000000  0.013612  0.108900  ...  1.259930  0.078931  0.005840   
4       0.176906  0.028280  0.000000  ...  1.100698  0.072530  0.029821   
...          ...       ...       ...  ...       ...       ...       ...   
247118  0.148386  0.214779  0.542384  ...  0.292787  0.000000  0.636535   
247119  0.000000  1.250417  0.036951  ...  1.123815  0.000000  0.097913   
247120  0.477553  0.000000  0.833972  ...  0.885310  0.000000  0.239653   
247121  0.000000  0.345184  0.066882  ...  0.314013  0.000000  0.023167   
247122  0.661369  0.129900  0.172631  ...  0.269158  0.000000  0.014829   

           V1275     V1276     V1277     V1278     V1279     V1280  \
0       0.002177  0.436013  0.378547  0.029962  0.127837  0.000000   
1       0.405357  0.068908  0.269205  0.000000  0.325799  0.036937   
2       0.544899  0.000000  0.000000  0.267199  0.880283  0.492699   
3       0.000000  0.011345  0.000000  0.000000  0.478731  0.518241   
4       0.139721  0.020297  0.000000  0.000000  0.019781  0.000000   
...          ...       ...       ...       ...       ...       ...   
247118  0.230335  0.000000  0.289987  1.239995  0.132540  0.270762   
247119  0.000000  0.000000  0.023646  0.102291  0.582427  0.024401   
247120  0.439827  0.004465  0.000000  0.048496  0.211974  0.088579   
247121  0.016230  0.000000  0.000000  0.000000  0.000000  0.147468   
247122  0.187695  0.000000  0.002154  0.486474  0.748830  0.622428   

               Color  
0       MULTICOLOR-6  
1       MULTICOLOR-6  
2       MULTICOLOR-6  
3       MULTICOLOR-6  
4       MULTICOLOR-6  
...              ...  
247118    MULTICOLOR  
247119    MULTICOLOR  
247120    MULTICOLOR  
247121    MULTICOLOR  
247122    MULTICOLOR  

[247123 rows x 1281 columns]

In [44]:
from sklearn.model_selection import train_test_split 
X = filtered_color_class.drop(['Color'], axis = 1)
y = filtered_color_class[['Color']]
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state=42)

In [45]:
from sklearn.linear_model import SGDClassifier
sgd_clf = SGDClassifier(max_iter=1000, tol=1e-3, random_state=42)

# Scaling in chunks 
batch_size = 10000
for start in range(0, len(X_train),batch_size):
    end = start + batch_size
    X_batch = X_train[start:end]
    y_batch = y_train[start:end]
    
    sgd_clf.partial_fit(X_batch,y_batch, classes = np.unique(y))

y_pred_1 = sgd_clf.predict(X_test)

c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was e

In [46]:
# Accuracy Testing 
from sklearn.metrics import accuracy_score, f1_score
# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_1))
print("F1 Score:", f1_score(y_test, y_pred_1, average='weighted'))

Accuracy: 0.2115528578654527
F1 Score: 0.23424255799961854


In [81]:

import os
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.preprocessing.image import img_to_array
from PIL import Image, UnidentifiedImageError
import requests
from io import BytesIO
import streamlit as st
import json
import datetime


def extract_feature_vector(image_path):
    # Load the pre-trained MobileNetV2 model
    base_model = MobileNetV2(weights='imagenet', include_top=False, pooling='avg')
    model = tf.keras.Model(inputs=base_model.input, outputs=base_model.output)

    # Open the image using PIL
    img = Image.open(image_path)

    if img.format != 'JPEG':
        # Convert image to RGB if it is not
        if img.mode != 'RGB':
            img = img.convert('RGB')
        
        img.save(image_path, format='JPEG')

    # Resize the image to the size expected by MobileNetV2 (224x224)
    img = img.resize((224, 224))
    # Convert the image to a numpy array and preprocess for the model
    img_array = np.array(img)
    img_array = tf.keras.applications.mobilenet_v2.preprocess_input(img_array)
    img_array = np.expand_dims(img_array, axis=0)  # Add batch dimension

    # Extract the feature vector using the model
    feature_vector = model.predict(img_array)

    return feature_vector

# Example usage
image_path = r"C:\Users\Admin\Attribute_Classification\Product_Attribute_Classification\Image_Vectorization\Image_24.jpeg"
feature_vector = extract_feature_vector(image_path)
image_vector_df = pd.DataFrame(feature_vector)
image_vector_df.columns = [f"V{i+1}" for i in range (image_vector_df.shape[1])]
image_vector_df 

1/1 [==============================] - 1s 752ms/step


V1        V2        V3   V4        V5        V6        V7   V8        V9  \
0  0.0  0.405876  0.064424  0.0  0.432766  0.299163  0.079397  0.0  0.230229   

   V10  ...  V1271     V1272     V1273  V1274     V1275     V1276     V1277  \
0  0.0  ...    0.0  0.105583  0.267501    0.0  1.360278  0.800593  0.033736   

      V1278  V1279     V1280  
0  0.271941    0.0  0.009499  

[1 rows x 1280 columns]

In [82]:
y_pred_1 = sgd_clf.predict(image_vector_df)
y_pred_1

array(['Boys - Kids'], dtype='<U12')

### Code for the MainCategory Classification

    > Training for the Maincategory

In [49]:
df_expanded

_id  VariantId   imageId  \
0       66e7803df37ca712f0413a2d    2307410  22147004   
1       66e7803df37ca712f0413a2e    2307410  22147005   
2       66e7803df37ca712f0413a2f    2307410  22147006   
3       66e7803df37ca712f0413a30    2307410  22147007   
4       66e7803df37ca712f0413a31    2307410  22147008   
...                          ...        ...       ...   
247118  66fa1eb37da136a03fe8a0fa    2643430  26680256   
247119  66fa1ea97da136a03fe8a0f7    2643431  26680257   
247120  66fa1ea97da136a03fe8a0f8    2643431  26680258   
247121  66fa1ea4928eb8beb6d3b6f0    2643432  26680259   
247122  66fa1ea4928eb8beb6d3b6f1    2643432  26680260   

                                                 imageUrl  \
0       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
1       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
2       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
3       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
4       https://shoekonnect.s3.amazonaws.com/PRODUCTIO...   
...                                                   ...   
247118  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247119  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247120  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247121  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   
247122  https://bijnis.s3.amazonaws.com/PRODUCTION/upl...   

                                                   vector  \
0       [0.02699310891330242, 0.026698676869273186, 0....   
1       [0.15216664969921112, 0.7531224489212036, 0.78...   
2       [0.27668434381484985, 0.0, 0.6610189080238342,...   
3       [0.0, 0.33198481798171997, 0.0, 0.0, 0.1082259...   
4       [0.015320106409490108, 0.13437797129154205, 0....   
...                                                   ...   
247118  [0.0, 0.0, 0.11995667964220047, 0.197625294327...   
247119  [0.12077472358942032, 0.0, 0.04706697165966034...   
247120  [0.0, 0.0, 0.5308622121810913, 0.0097048813477...   
247121  [0.015275663696229458, 0.0, 0.0472753793001174...   
247122  [0.0, 0.0, 0.015623705461621284, 0.01281377207...   

                  createdOn         Color ColorHex  Category MainCategory  \
0       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
1       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
2       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
3       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
4       2024-09-16 06:17:57  MULTICOLOR-6           t-shirts          Men   
...                     ...           ...      ...       ...          ...   
247118  2024-09-30 09:14:51    MULTICOLOR  #999999     jeans          Men   
247119  2024-09-30 09:14:41    MULTICOLOR  #999999     jeans          Men   
247120  2024-09-30 09:14:41    MULTICOLOR  #999999     jeans          Men   
247121  2024-09-30 09:14:36    MULTICOLOR  #999999     jeans          Men   
247122  2024-09-30 09:14:36    MULTICOLOR  #999999     jeans          Men   

        ...     V1271     V1272     V1273     V1274     V1275     V1276  \
0       ...  0.000000  0.502343  0.000000  0.099487  0.002177  0.436013   
1       ...  0.000000  0.839816  0.434324  0.000000  0.405357  0.068908   
2       ...  0.000000  0.000000  0.003457  0.100403  0.544899  0.000000   
3       ...  0.000000  1.259930  0.078931  0.005840  0.000000  0.011345   
4       ...  0.000000  1.100698  0.072530  0.029821  0.139721  0.020297   
...     ...       ...       ...       ...       ...       ...       ...   
247118  ...  0.015857  0.292787  0.000000  0.636535  0.230335  0.000000   
247119  ...  0.000000  1.123815  0.000000  0.097913  0.000000  0.000000   
247120  ...  0.000000  0.885310  0.000000  0.239653  0.439827  0.004465   
247121  ...  0.000000  0.314013  0.000000  0.023167  0.016230  0.000000   
247122  ...  0.000000  0.269158  0.000000  0.014829  0.187695  0.000000   

           V1277     V1278     V1279     V1280  
0  

In [50]:
column_names = df_expanded.columns.unique()
column_names

Index(['_id', 'VariantId', 'imageId', 'imageUrl', 'vector', 'createdOn',
       'Color', 'ColorHex', 'Category', 'MainCategory',
       ...
       'V1271', 'V1272', 'V1273', 'V1274', 'V1275', 'V1276', 'V1277', 'V1278',
       'V1279', 'V1280'],
      dtype='object', length=2278)

In [51]:
for i in column_names:
    if df_expanded[i].dtype == 'float64':
        df_expanded[i] = df_expanded[i].astype('float32')


In [52]:
filtered_columns = [col for col in df_expanded.columns if col.startswith('V') and col[1:].isdigit()] + ['MainCategory']
filtered_color_class = df_expanded[filtered_columns]
filtered_color_class

V1        V2        V3        V4        V5        V6        V7  \
0       0.026993  0.026699  0.013406  0.158621  0.000000  0.047574  1.477892   
1       0.152167  0.753122  0.780756  0.033834  0.224728  0.538430  0.347718   
2       0.276684  0.000000  0.661019  0.000000  0.000000  0.006237  0.270860   
3       0.000000  0.331985  0.000000  0.000000  0.108226  0.138597  1.015925   
4       0.015320  0.134378  0.000000  0.000000  0.001563  0.295301  0.045144   
...          ...       ...       ...       ...       ...       ...       ...   
247118  0.000000  0.000000  0.119957  0.197625  0.000000  0.000000  0.213593   
247119  0.120775  0.000000  0.047067  0.145077  0.000000  0.294627  0.264174   
247120  0.000000  0.000000  0.530862  0.009705  0.000000  0.000000  0.130015   
247121  0.015276  0.000000  0.047275  0.000000  0.020218  0.047031  0.177078   
247122  0.000000  0.000000  0.015624  0.012814  0.000000  0.000000  0.331728   

              V8        V9       V10  ...     V1272     V1273     V1274  \
0       0.136192  0.215258  0.034348  ...  0.502343  0.000000  0.099487   
1       0.000000  0.061816  0.000000  ...  0.839816  0.434324  0.000000   
2       0.095041  0.000000  0.736817  ...  0.000000  0.003457  0.100403   
3       0.000000  0.013612  0.108900  ...  1.259930  0.078931  0.005840   
4       0.176906  0.028280  0.000000  ...  1.100698  0.072530  0.029821   
...          ...       ...       ...  ...       ...       ...       ...   
247118  0.148386  0.214779  0.542384  ...  0.292787  0.000000  0.636535   
247119  0.000000  1.250417  0.036951  ...  1.123815  0.000000  0.097913   
247120  0.477553  0.000000  0.833972  ...  0.885310  0.000000  0.239653   
247121  0.000000  0.345184  0.066882  ...  0.314013  0.000000  0.023167   
247122  0.661369  0.129900  0.172631  ...  0.269158  0.000000  0.014829   

           V1275     V1276     V1277     V1278     V1279     V1280  \
0       0.002177  0.436013  0.378547  0.029962  0.127837  0.000000   
1       0.405357  0.068908  0.269205  0.000000  0.325799  0.036937   
2       0.544899  0.000000  0.000000  0.267199  0.880283  0.492699   
3       0.000000  0.011345  0.000000  0.000000  0.478731  0.518241   
4       0.139721  0.020297  0.000000  0.000000  0.019781  0.000000   
...          ...       ...       ...       ...       ...       ...   
247118  0.230335  0.000000  0.289987  1.239995  0.132540  0.270762   
247119  0.000000  0.000000  0.023646  0.102291  0.582427  0.024401   
247120  0.439827  0.004465  0.000000  0.048496  0.211974  0.088579   
247121  0.016230  0.000000  0.000000  0.000000  0.000000  0.147468   
247122  0.187695  0.000000  0.002154  0.486474  0.748830  0.622428   

        MainCategory  
0                Men  
1                Men  
2                Men  
3                Men  
4                Men  
...              ...  
247118           Men  
247119           Men  
247120           Men  
247121           Men  
247122           Men  

[247123 rows x 1281 columns]

In [53]:
from sklearn.model_selection import train_test_split 
X = filtered_color_class.drop(['MainCategory'], axis = 1)
y = filtered_color_class[['MainCategory']]
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state=42)

In [54]:
from sklearn.linear_model import SGDClassifier
sgd_clf = SGDClassifier(max_iter=1000, tol=1e-3, random_state=42)

# Scaling in chunks 
batch_size = 10000
for start in range(0, len(X_train),batch_size):
    end = start + batch_size
    X_batch = X_train[start:end]
    y_batch = y_train[start:end]
    
    sgd_clf.partial_fit(X_batch,y_batch, classes = np.unique(y))
y_pred_1 = sgd_clf.predict(X_test)

c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\Admin\miniconda3\envs\my_env\lib\site-packages\sklearn\utils\validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was e

In [55]:
# Accuracy Testing 
from sklearn.metrics import accuracy_score, f1_score
# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred_1))
print("F1 Score:", f1_score(y_test, y_pred_1, average='weighted'))

Accuracy: 0.8832776934749621
F1 Score: 0.8750028028866212


In [85]:
import os
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.preprocessing.image import img_to_array
from PIL import Image, UnidentifiedImageError
import requests
from io import BytesIO
import streamlit as st
import json
import datetime


def extract_feature_vector(image_path):
    # Load the pre-trained MobileNetV2 model
    base_model = MobileNetV2(weights='imagenet', include_top=False, pooling='avg')
    model = tf.keras.Model(inputs=base_model.input, outputs=base_model.output)

    # Open the image using PIL
    img = Image.open(image_path)

    if img.format != 'JPEG':
        # Convert image to RGB if it is not
        if img.mode != 'RGB':
            img = img.convert('RGB')
        
        img.save(image_path, format='JPEG')

    # Resize the image to the size expected by MobileNetV2 (224x224)
    img = img.resize((224, 224))
    # Convert the image to a numpy array and preprocess for the model
    img_array = np.array(img)
    img_array = tf.keras.applications.mobilenet_v2.preprocess_input(img_array)
    img_array = np.expand_dims(img_array, axis=0)  # Add batch dimension

    # Extract the feature vector using the model
    feature_vector = model.predict(img_array)

    return feature_vector

# Example usage
image_path = r"C:\Users\Admin\Attribute_Classification\Product_Attribute_Classification\Image_Vectorization\Image_21.jpeg"
feature_vector = extract_feature_vector(image_path)
image_vector_df = pd.DataFrame(feature_vector)
image_vector_df.columns = [f"V{i+1}" for i in range (image_vector_df.shape[1])]
image_vector_df 

1/1 [==============================] - 1s 902ms/step


V1        V2        V3        V4       V5        V6        V7   V8  \
0  0.0  1.604631  0.648798  0.296774  0.17829  0.081042  0.450518  0.0   

         V9  V10  ...  V1271     V1272  V1273    V1274  V1275     V1276  \
0  0.036803  0.0  ...    0.0  0.838832    0.0  0.15488    0.0  0.018348   

     V1277     V1278     V1279     V1280  
0  0.27358  0.333667  0.168951  0.231716  

[1 rows x 1280 columns]

In [86]:
y_pred_1 = sgd_clf.predict(image_vector_df)
y_pred_1

array(['Men'], dtype='<U12')

#### Guidelines to achieve the desired outcome
+ Pull the desired data from the MongoDB and MySQL DataBase 

        > Fetching Data from MongoDB is a blocker in training the model for multiple subcategories as it taking huge time due to no mapping of the SubCategories/MainCategories/SuperCategories 
        

+ Create a flat file with all the parameters and product attributes for further processing
+ Encode the whole data for the parameters to get the binary values of the output labels
+ Transform the vectors for 1280 (size of the output vector from MobileNet_V2) labels in sequential columns as an input
+ Identify all the relevant parameters/labels required to be identified using the input vectors 

    > P_2 Priority
    
+ Split the whole data into training, testing and validation set in (70-20-10 proportion)
    > Check for the feasibility of dimensionality reduction so that only relevant attributes gets into the model which will further help in efficiency and optimization of the model
    
    > Try to build an staged model which will filter Supercategory -> Subcategory -> Maincategory at first level, Color/Hex Code at Second Level and further attributes at final level

        > SubCategory Classification --> Done

        > Color Classification 

        > Attributes Classification

+ Train the model using two three major classifiers like SVC, Random Forest Classifier, tree based models like LightGBM, XGBoost and Neural Network etc.
+ Setup the MobileNet_V2 with weights = imagenet for the feature extraction from the input image

        > Status: Done
        
+ Get the output vector basis the model and flatten the same for the validation
+ Test for the accuracy, efficiency and optimization using advanced techniques and required metrics
+ Bundle the whole model and create an MVP using Streamlit Application for the realtime testing purpose